File: 104 - Maximum Depth of Binary Tree/cpp/ans.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // method 1 -> recursive
    int maxDepth(TreeNode* root) {
        if (root == nullptr) {
            // stop guard
            return 0;
        }
        return 1 + max(maxDepth(root->left), maxDepth(root->right));
    }
    // method 2 -> iteration
    int maxDepthIter(TreeNode* root) {
        if (root == nullptr) {
            // stop guard
            return 0;
        }
        int max_value = 0;
        vector<TreeNode*> v_node;
        vector<int> v_int;
        v_node.push_back(root);
        v_int.push_back(1);
        while (!v_node.empty()) {
            TreeNode* n = v_node.back();
            int depth = v_int.back();
            v_node.pop_back();
            v_int.pop_back();
            max_value = max(max_value, depth);

            if (n->left != nullptr) {
                v_node.push_back(n->left);
                v_int.push_back(depth + 1);
            }

            if (n->right != nullptr) {
                v_node.push_back(n->right);
                v_int.push_back(depth + 1);
            }
        }

        return max_value;
    }
};

File: 104 - Maximum Depth of Binary Tree/go/ans.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    // No built in Max :|
    left := maxDepth(root.Left)
    right := maxDepth(root.Right)
    if left > right {
        return 1 + left
    } else {
        return 1 + right
    }
}

func maxDepthIter(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    type Pair struct {
        Node *TreeNode
        Depth int
    }

    maxValue := 0
    stack := []Pair{Pair{Node: root, Depth: 1}}
    // loop through the stack
    for len(stack) > 0 {
        top := len(stack)-1
        curr := stack[top]  // get last element
        stack = stack[:top] // remove last element

        if maxValue < curr.Depth {
            maxValue = curr.Depth
        }

        if curr.Node.Left != nil {
            stack = append(stack, Pair{Node: curr.Node.Left, Depth: curr.Depth + 1})
        }
        if curr.Node.Right != nil {
            stack = append(stack, Pair{Node: curr.Node.Right, Depth: curr.Depth + 1})
        }
    }

    return maxValue
}